File: examples/hello.c
/*
 * hello.c - Simple example for native function tracing
 *
 * Build:   make
 * Trace:   ../malwi x -c malloc -c free ./hello
 * Stack:   ../malwi x -t -c malloc ./hello
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* greet(const char* name) {
    size_t len = strlen(name) + 10;
    char* msg = malloc(len);
    snprintf(msg, len, "Hello, %s!", name);
    return msg;
}

int main(int argc, char* argv[]) {
    const char* name = argc > 1 ? argv[1] : "World";

    char* message = greet(name);
    printf("%s\n", message);
    free(message);

    return 0;
}

File: examples/hello.py
#!/usr/bin/env python3
"""
hello.py - Simple example for Python function tracing

Trace:   ../malwi x -c py:greet python3 hello.py
Stack:   ../malwi x -t -c py:greet python3 hello.py
All:     ../malwi x -c 'py:*' python3 hello.py
"""

import sys


def greet(name: str) -> str:
    """Generate a greeting message."""
    return format_message(name)


def format_message(name: str) -> str:
    """Format the greeting."""
    return f"Hello, {name}!"


def main():
    name = sys.argv[1] if len(sys.argv) > 1 else "World"
    message = greet(name)
    print(message)


if __name__ == "__main__":
    main()
